### PR TITLE
feat: fetch all native stakers w/ history

### PR DESCRIPTION
**Pull request by butonium**

### CODE REVIEW

File: src/stakers/stakers.controller.ts [hunk 1]
```diff
@@ -22,6 +23,46 @@ import { StakersService } from './stakers.service';
...
+          HttpStatus.BAD_REQUEST,
+        );
+      }
+    }
+    if (!query.startDate && !query.endDate) {
+      throw new HttpException(
+        'No startDate or endDate provided',
+        HttpStatus.BAD_REQUEST,
+      );
+    }
```
Comment (martinkrecek) [inline src/stakers/stakers.controller.ts:54]: (non-blocking/optional comment) This is an exact copy-paste of the validation in `getNativeStakes` below, I'd consider wrapping it in a helper function which accepts the `query` as a parameter and performs the checks.

File: src/stakers/stakers.controller.ts [hunk 2]
```diff
@@ -22,6 +23,46 @@ import { StakersService } from './stakers.service';
...
 
+
+  @Get('/ns/all')
+  @ApiOperation({
+    summary: 'Fetch Native Stake balances for a specific date interval',
+  })
+  @ApiResponse({
+    status: 200,
+    description: 'The records were successfully fetched.',
+    type: NativeStakeBalanceDto,
```
Comment (martinkrecek) [inline src/stakers/stakers.controller.ts:34]: Using `NativeStakeBalanceDto` and `NativeStakeBalancesDto` can be dangerous exactly for these reasons, this line is supposed to contain `NativeStakeBalancesDto` (the newly added DTO). This line should definitely be fixed, but I'd also recommend renaming the new DTO to something like `NativeStakeAllBalancesDto` or `AllNativeStakeBalancesDto` to differentiate it a bit more.

File: src/stakers/stakers.service.ts [hunk 3]
```diff
@@ -44,4 +44,47 @@ export class StakersService {
     this.logger.log('Staker data fetched', { withdraw_authority, result });
     return balances;
   }
+
+  async getAllNativeStakeBalances(
+    startDate: string,
+    endDate: string,
+  ): Promise<{ [key: string]: NativeStakeBalanceDto[] } | null> {
```
Comment (martinkrecek) [inline src/stakers/stakers.service.ts:51]:
```suggestion
  ): Promise<NativeStakeBalancesDto | null> {
```

Haven't tested it, but shouldn't this work too?

File: src/stakers/stakers.service.ts [hunk 4]
```diff
@@ -44,4 +44,47 @@ export class StakersService {
...
+        LEFT JOIN snapshots ON snapshots.snapshot_id = native_stake_accounts.snapshot_id
+        WHERE snapshots.created_at >= ${startDate} AND snapshots.created_at <= ${endDate}
+      `);
+
+    if (!result || result.length === 0) {
+      this.logger.warn('Stakes not found!');
+      return null;
+    }
+
+    const ownerBalances: { [key: string]: NativeStakeBalanceDto[] } = {};
```
Comment (martinkrecek) [inline src/stakers/stakers.service.ts:72]: Same case like in the comment on the function signature above, using `NativeStakeBalancesDto` should be ok here.

**Review by martinkrecek — CHANGES_REQUESTED** [review]
Commenting mostly on data types, otherwise looks ok.

**Review by martinkrecek — APPROVED** [review]

**Review by janlegner — APPROVED** [review]

**Pull request merged**